### DISCUSSION TITLE
Advanced Hyperparameter Tuning for Time Series Models in PyCaret

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1795] [FAQs]
# Overview 

In **[this FAQ](https://github.com/pycaret/pycaret/discussions/1791)**, we saw how to perform hyper-parameter tuning for time series with just a few lines of code. We also explored the search space used for tuning which gave us better insight to what is happening under the hood in `pycaret`. Before delving deeper into this advanced FAQ, please read the previous linked FAQ first as the advanced FAQ builds upon it.

While the default tuning works well in many cases, sometimes the user might want to take a more hands on approach and provide their own custom search space. In this post, we will explore how the user can

1. **Evaluate** the performance of the default `pycaret` search space
2. **Find a custom (alternate) search space** that is more in line with the needs of their specific data set, and
2. **Use this custom search space** for tuning the models in `pycaret`

An overview notebook that shows these steps can be **[found here](https://gist.github.com/ngupta23/5ea07c429625ed536aa7922e19e94ec4)**.

## Step 1: Evaluating the performance of the default `pycaret` search space

```python
#### What is the default search space (random grid search)? ----
random_grid = exp.models(internal=True).loc["lr_cds_dt", "Tune Distributions"]
random_grid
```

![image](https://user-images.githubusercontent.com/33585645/140041968-580ed246-f4be-49c4-9eec-bc3f12f7ad2b.png)

```python
#### Create a baseline model ----
model = exp.create_model("lr_cds_dt")

#### Tune the model. How did the best model perform?  ----
# NOTE: we set `return_tuner=True` since we want to check the performance of the search space later
tuned_model_random, random_tuner = exp.tune_model(model, return_tuner=True)
```

![image](https://user-images.githubusercontent.com/33585645/140043225-9f8b5989-4859-4975-b146-b2bb634e90c7.png)

**Observations**

1. The best model with the default random search space gives a mean cross validation sMAPE = 0.0452 (lower is better)

```python
#### How did the entire search space perform and how were the best hyper-parameters selected? ----
# NOTE: We can use the returned tuner to check the performance of the entire search space.
import pandas as pd
pd.DataFrame(random_tuner.cv_results_).sort_values("rank_test_smape")
```

![image](https://user-images.githubusercontent.com/33585645/140042631-aee943ba-1c33-4c34-af68-d41894b1f6e1.png)

## Step 2: Finding custom search space in line with specific data set needs

**Observations from search space exploration from Step 1**

1.   Seasonal period of harmonics of 12 (e.g. 24) can be helpful (top 2 hyper-parameter combinations)
2.   Lower de-trending degree (2, 3) gives better results compared to higher values

So why not limit the search space using these observations to see if we can improve the performance more? Let's do that next.

## Step 3: Tuning with custom search space

Based on the observations in Step 2

1. lets add some more harmonics for the seasonal period. We will use 12, 24, and 36, and
2. lets limit the de-trending degree to 1, 2, 3

Rest of the hyper-parameter can remain the same as the ones in the default search space. 

```python
#### Update the search space ----
random_grid['sp'] = [12, 24, 36]
random_grid['degree'] = [1, 2, 3]
random_grid
```

![image](https://user-images.githubusercontent.com/33585645/140043130-65cd622b-700c-470e-bc7a-b032b9e646f4.png)

In addition, the tuning was very quick, so we can afford to try more combinations. We will increase the number of searches from 10 (default) to 20 for this run.

```python
better_tuned_random = exp.tune_model(model, custom_grid=random_grid, n_iter=20)
```

![image](https://user-images.githubusercontent.com/33585645/140044450-e9a4d333-1a9b-4b41-b737-689433cdd996.png)

**Observations**

- Model performance during cross-validation has improved. For example, sMAPE reduced from 0.0452 (in the default random grid search case) to 0.0391using the custom grid.

That is it for this FAQ. Happy forecasting!

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
